### PR TITLE
Fix: Make resource syncs respect base dir option

### PR DESCRIPTION
**Pull request by qingchm**
#### Which issue(s) does this change fix?
N.A.

#### Why is this change necessary?
Without this change any definition files will not be found correctly if base_dir option is not respected for customers who have code directory somewhere else

#### How does it address the issue?
It adds the base_dir into the full path of the definition files

#### What side effects does this change have?
N.A.

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by CoshUS — CHANGES_REQUESTED** [review]
nit: Can we add description in the PR for future tracking?
Rest looks good to me. Thanks for addressing this!

**Comment by super132** [comment]
> #### Which issue(s) does this change fix?
> #### Why is this change necessary?
> #### How does it address the issue?
> #### What side effects does this change have?
> #### Checklist
> * [ ]  Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
> * [ ]  Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
> * [x]  Write unit tests
> * [ ]  Write/update functional tests
> * [ ]  Write/update integration tests
> * [x]  `make pr` passes
> * [ ]  `make update-reproducible-reqs` if dependencies were changed
> * [ ]  Write documentation
> 
> By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

Can you please link the issue that this PR addresses and a description of why do we need this change and what is the change about?

Thanks.

File: tests/unit/lib/sync/flows/test_rest_api_sync_flow.py [hunk 1]
```diff
@@ -275,6 +278,16 @@ def test_get_definition_file(self, get_resource_mock):
...
 
+    @patch("samcli.lib.sync.flows.generic_api_sync_flow.get_resource_by_id")
+    def test_get_definition_file_with_base_dir(self, get_resource_mock):
+        sync_flow = self.create_sync_flow()
+
+        sync_flow._build_context.base_dir = "base_dir"
+        get_resource_mock.return_value = {"Properties": {"DefinitionUri": "test_uri"}}
+        result_uri = sync_flow._get_definition_file("test")
+
+        self.assertEqual(result_uri, "base_dir/test_uri")
```
Comment (CoshUS) [inline tests/unit/lib/sync/flows/test_rest_api_sync_flow.py:289]: "/" is for unix. On Windows "\" must be used instead.
Use "os.path.sep" instead.

File: tests/unit/lib/sync/flows/test_stepfunctions_sync_flow.py [hunk 2]
```diff
@@ -70,6 +70,16 @@ def test_get_definition_file(self, join_path_mock, get_resource_mock):
...
 
+    @patch("samcli.lib.sync.flows.stepfunctions_sync_flow.get_resource_by_id")
+    def test_get_definition_file_with_base_dir(self, get_resource_mock):
+        sync_flow = self.create_sync_flow()
+
+        sync_flow._build_context.base_dir = "base_dir"
+        sync_flow._resource = {"Properties": {"DefinitionUri": "test_uri"}}
+        result_uri = sync_flow._get_definition_file("test")
+
+        self.assertEqual(result_uri, "base_dir/test_uri")
```
Comment (mndeveci) [inline tests/unit/lib/sync/flows/test_stepfunctions_sync_flow.py:81]: Please replace this forward slash as well.

**Review by mndeveci — APPROVED** [review]
Thanks for working on this, it looks good other than the feedback that @CoshUS gave.

**Review by CoshUS — APPROVED** [review]

**Pull request merged**